vision/odlc_contour_filtering.py

```diff
@@ -0,0 +1,407 @@
+    have relatively smooth sides, and the approximated contour will not have a lot of change. If
+    the contour is not actually a shape (ie a patch of grass), then its sides will be less
+    straight, and the polygon approximation will be very different from the original.
+
+    parameters
+    ----------
+    contour : contour_type
+        The originally found contour to be evaluated (as returned from cv2.findContours)
+    approx : contour_type
+        The contour (same as contour param) but run through an approximation algoritm
+        such as cv2.approxPolyDP(contour, 0.05*cv2.arcLength(contour, True), True)
+                                          ^^^^can be tweaked
+    test_percent_diff : float
+        The max percent difference in area allowed between the original and approximaed contour
+
+    returns
+    -------
+    returns true if the non-overlapping area between the two contours is less than the specified
+    amount, false otherwise
+    """
+    # kinda need to figure out how much non overlap is too much
+    box: bound_box_type = _min_common_bounding_box((contour, approx))
+
+    contour_mask: mask_type = _generate_mask(contour, box)
+    approx_mask: mask_type = _generate_mask(approx, box)
+
+    non_overlap_mask: mask_type = np.logical_xor(contour_mask, approx_mask)
+
+    non_overlap_cnts: tuple[contour_type]
+    # non_overlap_hier: hierarchy_type
+    non_overlap_img: sc_image_type = non_overlap_mask.astype(UInt8)
+    non_overlap_cnts, _ = cv2.findContours(non_overlap_img, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+
+    non_overlap_area_sum = 0
+    for cnt in non_overlap_cnts:
+        non_overlap_area_sum += cv2.contourArea(cnt)
+
+    contour_area = cv2.contourArea(contour)
+    if (
+        len(non_overlap_cnts) == 0
+        or non_overlap_area_sum == 0
+        or 1 - test_percent_diff < contour_area / non_overlap_area_sum < 1 + test_percent_diff
+    ):
+        return True
+    return False
+
+
+def test_circleness(img: sc_image_type) -> bool:
+    """
+    this function will test the cropped area around a given contour to see if it is a circle or if
+    it is not a circle, it uses the cv2.HoughCircles() function to check for a circle
+
+    parameters
+    ----------
+    img : sc_image_type
+        img should be grayscale cropped box around contour and PROCESSED ALREADY
+        as in the image is binary with the contour filled in as white
+        NOTE: sc_image_type denotes a single channel image
+
+    returns
+    -------
+    returns true if a circle of appropriate size is found (to reduce chance of false positives)
+    """
+    padding: int = int(img.shape[0] * 0.05)
+    modded: sc_image_type = cv2.copyMakeBorder(
+        img, padding, padding, padding, padding, cv2.BORDER_CONSTANT, None, 0
+    )
+    modded = cv2.GaussianBlur(img, (5, 5), 5)
+    circles: circles_type | None = cv2.HoughCircles(
+        modded,
```

Comment:
Does circles need its own type if it's only used here? Same thing for some of the other types. If they are only used once, they don't really need a type alias. Make sure you re-evaluate which ones actually need to have an alias.